crates/filtering/src/madgwick.rs

```diff
@@ -0,0 +1,94 @@
+use nalgebra::{
+    Matrix4, Quaternion, RealField, Scalar, SimdValue, UnitQuaternion, Vector3, Vector4,
+};
+use num_traits::{One, Zero};
+
+#[derive(Debug)]
+pub struct AccelerometerNormZero;
+
+/// Madgwick's filter for orientation estimation.
+///
+/// This implementation is based on the following:
+/// https://github.com/jmagnuson/ahrs-rs
+pub trait Madgwick<T> {
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero>;
+
+    fn update_with_gyro(&mut self, gyroscope: &Vector3<T>, sample_period: T);
+}
+
+impl<T> Madgwick<T> for UnitQuaternion<T>
+where
+    T: Scalar + SimdValue + RealField + One + Zero + Copy,
+{
+    #[allow(non_snake_case)]
+    fn update_with_imu(
+        &mut self,
+        gyroscope: &Vector3<T>,
+        accelerometer: &Vector3<T>,
+        filter_gain: T,
+        sample_period: T,
+    ) -> Result<(), AccelerometerNormZero> {
+        let q = self.as_ref();
+
+        let zero: T = nalgebra::zero();
+        let two: T = nalgebra::convert(2.0);
+        let four: T = nalgebra::convert(4.0);
+        let half: T = nalgebra::convert(0.5);
+
+        // Normalize accelerometer measurement
+        let Some(accel) = accelerometer.try_normalize(zero) else {
+            return Err(AccelerometerNormZero);
+        };
+
+        // Gradient descent algorithm corrective step
+        #[rustfmt::skip]
+        let F = Vector4::new(
+            two*(       q[0]*q[2] - q[3]*q[1]) - accel[0],
```

Comment:
please prefer `i,j,k,w` indexing to `[0], [1], [2], [3]`

Reply:
done